### PR TITLE
feat: rework snapshots and add tests

### PR DESCRIPTION
**Pull request by BrechtDeVlieger**
### Description

Initially, this PR was meant to fix the broken `invalidate_hard_deletes` functionality of the glue table snapshots. However, fixing this was pretty much impossible, because of how the snapshots were implemented.

The current implementation of the snapshots for non-iceberg tables differs from dbt-core. The result has different column names and values. This PR uses as many of the original macros as possible to assure that the output is the same.

**Note: these changes are not backward incompatible.** Any existing use of snapshots will break because of this. But I strongly believe that this is a necessary change. To make the migration easier, I added some logic that detects when the invalid structure was used. In that case, we print an error message explaining the backward incompatible changes. The error also prints a set of queries that can be used to migrate the snapshot table. 

An example error message looks as follows:

```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
The snapshot logic of dbt-athena has changed in an incompatible way to be more consistent with the dbt-core implementation.
You will need to migrate your existing snapshot tables to be able to keep using them with the latest dbt-athena version.
You can find more information in the release notes:
https://github.com/dbt-athena/dbt-athena/releases
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

You can use the example query below as a baseline to perform the migration:

------------------------------------------------------------------------------------------
-- Backup original table
create table "awsdatacatalog"."dev_brechtdevliegershowpadcom_gold_foundations"."cc_date_snapshot__dbt_tmp_migration_backup"
  with (
    table_type='hive',
    is_external=true,external_location='s3://aws-athena-query-results-eu-central-1-549731708918/dbt/dbt-dev/dbt-brechtdevliegershowpadcom/tables/dev_brechtdevliegershowpadcom_gold_foundations/cc_date_snapshot__dbt_tmp_migration_backup/8914af21-55c4-4dce-aaf7-ac1284748914',
    format='parquet'
  )
  as
    select * from "awsdatacatalog"."dev_brechtdevliegershowpadcom_gold_foundations"."cc_date_snapshot";


-- Store new results in staging table
create table "awsdatacatalog"."dev_brechtdevliegershowpadcom_gold_foundations"."cc_date_snapshot__dbt_tmp_migration_staging"
  with (
    table_type='hive',
    is_external=true,external_location='s3://aws-athena-query-results-eu-central-1-549731708918/dbt/dbt-dev/dbt-brechtdevliegershowpadcom/tables/dev_brechtdevliegershowpadcom_gold_foundations/cc_date_snapshot__dbt_tmp_migration_staging/311ba50a-5a26-4607-bf48-8ea73ea077de',
    format='parquet'
  )
  as
    select
    id,
    name,
    some_date,
    dbt_snapshot_at as dbt_updated_at,
    dbt_valid_from,
    if(dbt_valid_to > cast('9000-01-01' as timestamp), null, dbt_valid_to) as dbt_valid_to,
    dbt_scd_id
from "awsdatacatalog"."dev_brechtdevliegershowpadcom_gold_foundations"."cc_date_snapshot"
where dbt_change_type != 'delete'
;


-- Drop target table
-- Note: you will need to manually remove the S3 files if you have a static table location

drop table `awsdatacatalog`.`dev_brechtdevliegershowpadcom_gold_foundations`.`cc_date_snapshot`;


-- Copy staging to target
create table "awsdatacatalog"."dev_brechtdevliegershowpadcom_gold_foundations"."cc_date_snapshot"
  with (
    table_type='hive',
    is_external=true,external_location='s3://aws-athena-query-results-eu-central-1-549731708918/dbt/dbt-dev/dbt-brechtdevliegershowpadcom/tables/dev_brechtdevliegershowpadcom_gold_foundations/cc_date_snapshot/d49a5ff5-80a4-413e-8dcb-281c212a1439',
    format='parquet'
  )
  as
    select * from "awsdatacatalog"."dev_brechtdevliegershowpadcom_gold_foundations"."cc_date_snapshot__dbt_tmp_migration_staging";


-- Drop staging table
drop table `awsdatacatalog`.`dev_brechtdevliegershowpadcom_gold_foundations`.`cc_date_snapshot__dbt_tmp_migration_staging`;
------------------------------------------------------------------------------------------
```

## Models used to test - Optional
I added functional tests for the snapshots, based on the original dbt-core tests. The tests run for both glue and iceberg tables.

I also tested the migration logic. For that I used the following models:

seeds/base.csv
```csv
id,name,some_date
1,Easton,1981-05-20T06:46:51
2,Lillian,1978-09-03T18:10:33
3,Jeremiah,1982-03-11T03:59:51
4,Nolan,1976-05-06T20:21:35
5,Hannah,1982-06-23T05:41:26
6,Eleanor,1991-08-10T23:12:21
7,Lily,1971-03-29T14:58:02
8,Jonathan,1988-02-26T02:55:24
9,Adrian,1994-02-09T13:14:23
10,Nora,1976-03-01T16:51:39
```

seeds/added.csv
```
id,name,some_date
4,Updated Nolan,2022-05-06T20:21:35
5,Updated Hannah,2022-06-23T05:41:26
6,Updated Eleanor,2022-08-10T23:12:21
7,Lily,1971-03-29T14:58:02
8,Jonathan,1988-02-26T02:55:24
9,Adrian,1994-02-09T13:14:23
10,Nora,1976-03-01T16:51:39
11,New Mateo,2014-09-07T17:04:27
12,New Julian,2000-02-04T11:48:30
13,New Gabriel,2001-07-10T07:32:52
```

snapshots/cc_all_snapshot.sql
```sql
{% snapshot cc_all_snapshot %}
    {{ config(
        check_cols='all', unique_key='id', strategy='check', table_type='hive'
    ) }}
    select * from {{ ref(var('seed_name', 'base')) }}
{% endsnapshot %}
```

snapshots/cc_date_snapshot.sql
```sql
{% snapshot cc_date_snapshot %}
    {{ config(
        check_cols=['some_date'], unique_key='id', strategy='check', table_type='hive'
    ) }}
    select * from {{ ref(var('seed_name', 'base')) }}
{% endsnapshot %}
```

snapshots/cc_name_snapshot.sql
```sql
{% snapshot cc_name_snapshot %}
    {{ config(
        check_cols=['name'], unique_key='id', strategy='check',
        table_type='hive'
    ) }}
    select * from {{ ref(var('seed_name', 'base')) }}
{% endsnapshot %}
```

snapshots/ts_snapshot.sql
```sql
{% snapshot ts_snapshot %}
    {{ config(
        strategy='timestamp',
        unique_key='id',
        updated_at='some_date',
        table_type='hive'
    )}}
    select * from {{ ref(var('seed_name', 'base')) }}
{% endsnapshot %}
```

> Note: I tried these snapshots with `invalidate_hard_deletes=True` as well, but that functionality is broken for all snapshot strategies

Steps to reproduce:
- Install the dbt-athena package from main
- Run `dbt seed`
- Run `dbt snapshot`
- Run `dbt snapshot --vars "seed_name: added"`
- Install dbt-athena from this branch
- Run `dbt snapshot --vars "seed_name: added"` again => Error
- Copy the provided queries and run them one by one in Athena
- Repeat this until all tables are migrated
- Run `dbt snapshot --vars "seed_name: added"` again => Success
- Success => Check contents of the tables

## Checklist
- [x] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [x] You added unit testing when necessary
- [x] You added functional testing when necessary

### CODE REVIEW

File: dbt/include/athena/macros/utils/timestamps.sql [hunk 1]
```diff
@@ -3,7 +3,18 @@
 %}
 
 {% macro athena__current_timestamp() -%}
-    cast(now() as timestamp(6))
+  {{ cast_timestamp('now()') }}
+{%- endmacro %}
+
+
+{% macro cast_timestamp(timestamp_col) -%}
+  {%- set config = model['config'] -%}
```
Comment (nicor88) [inline dbt/include/athena/macros/utils/timestamps.sql:11]: smart idea. 

What if we are using this in a context without config?

**Comment by nicor88** [comment]
this pr should address this issue: https://github.com/dbt-athena/dbt-athena/issues/197

File: dbt/include/athena/macros/utils/timestamps.sql (hunk 1 above)
Comment (BrechtDeVlieger) [reply dbt/include/athena/macros/utils/timestamps.sql:11]: Good point, do you know if the `model` variable is always available? If so, I can change it to 
```
{%- set config = model.get('config', {}) -%}
```

Comment (nicor88) [reply dbt/include/athena/macros/utils/timestamps.sql:11]: you can try, I believe that for tests is not available.

Comment (BrechtDeVlieger) [reply dbt/include/athena/macros/utils/timestamps.sql:11]: Ah right, you meant dbt tests of course. I checked that, and it works as expected:

```
SELECT {{ cast_timestamp('now()') }}
```
compiles to 
```
SELECT cast(now() as timestamp)
```

And 
```
{{ config(table_type='iceberg') }}

SELECT {{ cast_timestamp('now()') }}
```
compiles to
```
SELECT cast(now() as timestamp(6))
```

This was done by adding the mentioned content to `tests/test_timestamp.py` and then running `dbt test --select test_timestamp`

Comment (BrechtDeVlieger) [reply dbt/include/athena/macros/utils/timestamps.sql:11]: So it works with and without config. By default it will use the glue version, which makes sense imo

Comment (nicor88) [reply dbt/include/athena/macros/utils/timestamps.sql:11]: Yes absolutely, thanks for checking

**Comment by Avinash-1394** [comment]
Can you provide examples of this working for both iceberg and hive tables? Preferably one for each strategy type check and timestamp

**Comment by BrechtDeVlieger** [comment]
In the functional tests you can find working examples for both glue and iceberg. Those are the snapshot tests in the test_basic.py tests file

File: dbt/include/athena/macros/adapters/relation.sql [hunk 2]
```diff
@@ -5,7 +5,11 @@
       {%- do adapter.clean_up_table(relation.schema, relation.table) -%}
     {%- endif %}
     {% call statement('drop_relation', auto_begin=False) -%}
-      drop {{ relation.type }} if exists {{ relation.render_hive() }}
+      {%- if relation.type == 'view' -%}
+        drop {{ relation.type }} if exists {{ relation.render() }}
+      {%- else -%}
+        drop {{ relation.type }} if exists {{ relation.render_hive() }}
+      {% endif %}
```
Comment (nicor88) [inline dbt/include/athena/macros/adapters/relation.sql:12]: this has nothing to do with snapshot logic, could you please open another pull request for that? we aim to have minimal changes to isolate where some issues came from. Thanks.

File: dbt/include/athena/macros/materializations/models/table/create_table_as.sql [hunk 3]
```diff
@@ -48,7 +48,7 @@
   with (
     table_type='{{ table_type }}',
     is_external={%- if table_type == 'iceberg' -%}false{%- else -%}true{%- endif %},
-  {%- if work_group_output_location is none -%}
+  {%- if table_type == 'iceberg' or work_group_output_location is none -%}
```
Comment (nicor88) [inline dbt/include/athena/macros/materializations/models/table/create_table_as.sql:51]: what is the reason behind this? what if people using a specified `work_group_output_location ` and iceberg want to use the default workgroup location? 
Also, this seems a bit unrelated to the pull request, eventually, consider to move it into another PR ;) https://github.com/dbt-athena/dbt-athena/pull/202/files#r1164195147

File: tests/functional/adapter/test_basic.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
+import pytest
```
Comment (nicor88) [inline tests/functional/adapter/test_basic.py:1]: nice testing, but agina as here: https://github.com/dbt-athena/dbt-athena/pull/202/files#r1164195147 please keep minimal changes, makes the review easier :)

**Comment by nicor88** [comment]
@BrechtDeVlieger before reviewing this PR consider to split your work in multiple Pull requests that cover different issues, contributions are really welcome, but "we" avoid to have Pull requests that cover multiple issues in one. 

Thanks.

**Review by nicor88 — CHANGES_REQUESTED** [review]
@BrechtDeVlieger before reviewing this PR consider splitting your work into multiple Pull requests that cover different issues, contributions are really welcome, but "we" avoid having Pull requests that cover various topics in one big pr.

Thanks.

File: dbt/include/athena/macros/adapters/relation.sql (hunk 2 above)
Comment (jessedobbelaere) [reply dbt/include/athena/macros/adapters/relation.sql]: Removed it for him 👍 (Brecht is 🌴 )

File: dbt/include/athena/macros/materializations/models/table/create_table_as.sql (hunk 3 above)
Comment (BrechtDeVlieger) [reply dbt/include/athena/macros/materializations/models/table/create_table_as.sql:51]: Right. I’ll split the PR next week. It’s a bit tricky though. The tests that I added are the basic test suite provided by dbt. All the changes I did were done in order to fix those tests. I could create a separate PR where I add the tests and include the fixes to make those tests work? I’d then skip the hive snapshot tests for now. Then the second PR would contain the changes related to the hive snapshots. Wdyt?

Comment (BrechtDeVlieger) [reply dbt/include/athena/macros/materializations/models/table/create_table_as.sql:51]: And related to this comment: this actually gave me troubles. I had a workgroup with a default location, dbt failed when it tried to create a table, because the ‘location’ keyword was skipped. Apparently iceberg tables take the default location from the schema, not the athena workgroup.

Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/table/create_table_as.sql:51]: Ah gotcha, make sense as they are managed tables :)
Regarding splitting I'm fine to keep all the basic test suite in this one, as it could create less trouble.

Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/table/create_table_as.sql:51]: This should be covered by https://github.com/dbt-athena/dbt-athena/pull/231

**Comment by nicor88** [comment]
@BrechtDeVlieger regarding the refactor of this pr and the split:
- most of the adatper fixes are being covered
- let's leave this as it is(including tests), just clean up the already covered issues like this https://github.com/dbt-athena/dbt-athena/pull/202#discussion_r1169630739

File: dbt/include/athena/macros/materializations/snapshots/snapshot.sql [hunk 5]
```diff
@@ -12,365 +11,112 @@
...
-
-        {%- set select = snapshot_staging_table(strategy, sql, target_relation, table_type) -%}
-
-        {% call statement('build_snapshot_staging_relation') %}
-            {{ create_table_as(False, tmp_relation, select) }}
-        {% endcall %}
-
-        {% do return(tmp_relation) %}
-    {% endif %}
+{% macro iceberg_snapshot_merge_sql(target, source, insert_cols) %}
```
Comment (BrechtDeVlieger) [inline dbt/include/athena/macros/materializations/snapshots/snapshot.sql:31]: The original `athena__snapshot_merge_sql` macro is split in two separate macros, one for hive and one for iceberg. This is the iceberg part of that function without any other changes.

File: dbt/include/athena/macros/materializations/snapshots/snapshot.sql [hunk 6]
```diff
@@ -12,365 +11,112 @@
...
-
-        {% set sql -%}
-          select * from {{ source }};
-        {%- endset -%}
-
-        {{ create_table_as(False, target_relation, sql) }}
-    {% endif %}
-{% endmacro %}
+#}
+{% macro hive_snapshot_merge_sql(target, source, insert_cols, table_type) -%}
```
Comment (BrechtDeVlieger) [inline dbt/include/athena/macros/materializations/snapshots/snapshot.sql:18]: The original `athena__snapshot_merge_sql` macro is split in two separate macros, one for hive and one for iceberg. This is the hive part of that function without any other changes.

File: dbt/include/athena/macros/materializations/snapshots/snapshot.sql (hunk 5 above)
Comment (BrechtDeVlieger) [reply dbt/include/athena/macros/materializations/snapshots/snapshot.sql:31]: It was not possible to use the original dbt-core merge sql, because Athena requires an alias for the source table.

File: dbt/include/athena/macros/materializations/snapshots/snapshot.sql [hunk 7]
```diff
@@ -12,365 +11,112 @@
...
-        select * from upserts;
-        {% endif %}
-    {% endif %}
-{%- endmacro %}
-
-{%
-  Create a new temporary table to hold the records to be upserted or deleted
-%}
-
-{% macro athena__build_snapshot_staging_table(strategy, sql, target_relation, table_type) %}
```
Comment (BrechtDeVlieger) [inline dbt/include/athena/macros/materializations/snapshots/snapshot.sql:271]: This macro is gone, because we now use the implementation from dbt-core.

File: dbt/include/athena/macros/materializations/snapshots/snapshot.sql [hunk 8]
```diff
@@ -379,9 +125,8 @@
   {%- set target_table = model.get('alias', model.get('name')) -%}
 
   {%- set strategy_name = config.get('strategy') -%}
-  {%- set unique_key = config.get('unique_key') -%}
   {%- set file_format = config.get('file_format', 'parquet') -%}
-  {%- set table_type = config.get('table_type', 'iceberg') -%}
```
Comment (BrechtDeVlieger) [inline dbt/include/athena/macros/materializations/snapshots/snapshot.sql:384]: I changed the default from `iceberg` to `hive`, to be consistent with the other defaults in dbt-athena

File: dbt/include/athena/macros/materializations/snapshots/snapshot.sql [hunk 9]
```diff
@@ -379,9 +125,8 @@
   {%- set target_table = model.get('alias', model.get('name')) -%}
 
   {%- set strategy_name = config.get('strategy') -%}
-  {%- set unique_key = config.get('unique_key') -%}
```
Comment (BrechtDeVlieger) [inline dbt/include/athena/macros/materializations/snapshots/snapshot.sql:382]: This was nowhere used.

File: dbt/include/athena/macros/materializations/snapshots/snapshot.sql [hunk 10]
```diff
@@ -409,15 +154,20 @@
...
 
-      {% set build_sql = build_snapshot_table(strategy, model['compiled_sql'], table_type) %}
+      {% set build_sql = build_snapshot_table(strategy, model['compiled_sql']) %}
       {% set final_sql = create_table_as(False, target_relation, build_sql) %}
 
   {% else %}
 
       {{ adapter.valid_snapshot_target(target_relation) }}
 
+      {% set staging_relation = make_temp_relation(target_relation) %}
```
Comment (BrechtDeVlieger) [inline dbt/include/athena/macros/materializations/snapshots/snapshot.sql:164]: This piece of logic removes the existing staging table in case it wasn't properly removed in the previous run.

File: tests/functional/adapter/test_snapshot.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
+"""
```
Comment (BrechtDeVlieger) [inline tests/functional/adapter/test_snapshot.py:1]: I had to alter the original tests, because UPDATE statements are not allowed on seed tables. So instead, I use different seeds for each stage of the test. We could try to get this changed in dbt-core as well, so that we can use the default test suite again?

File: dbt/include/athena/macros/materializations/snapshots/snapshot.sql [hunk 12]
```diff
@@ -12,365 +11,112 @@
...
-{% macro athena__create_new_snapshot_table(strategy, strategy_name, target, source) %}
-    {%- set tmp_identifier = target.identifier ~ '__dbt_tmp_1' -%}
-
-    {%- set tmp_relation = adapter.get_relation(database=target.database, schema=target.schema, identifier=tmp_identifier) -%}
-
-    {%- set target_relation = api.Relation.create(identifier=tmp_identifier,
-      schema=target.schema,
-      database=target.database,
-      type='table') -%}
+{% macro hive_create_new_snapshot_table(target, source, insert_cols) %}
```
Comment (BrechtDeVlieger) [inline dbt/include/athena/macros/materializations/snapshots/snapshot.sql:60]: This is actually the major change of this PR. This is some sql logic to replicate the `MERGE INTO` statements of dbt-core. The output should be exactly the same as the MERGE INTO output, with the main difference that the results are first stored in a temporary staging table, and only then moved to the final location.

File: dbt/adapters/athena/impl.py [hunk 13]
```diff
@@ -731,3 +732,97 @@ def get_columns_in_relation(self, relation: AthenaRelation) -> List[AthenaColumn
...
+                "with the dbt-core implementation.\nYou will need to migrate your existing snapshot tables to be "
+                "able to keep using them with the latest dbt-athena version.\nYou can find more information "
+                "in the release notes:\nhttps://github.com/dbt-athena/dbt-athena/releases\n"
+                f"{'!'*90}\n\n"
+                "You can use the example query below as a baseline to perform the migration:\n\n"
+                f"{'-'*90}\n"
+                f"{sql}\n"
+                f"{'-'*90}\n\n"
+            )
+            print(msg)
```
Comment (Jrmyy) [inline dbt/adapters/athena/impl.py:759]: Can you remove this or put it as logging debug if needed ?

File: dbt/adapters/athena/impl.py [hunk 14]
```diff
@@ -731,3 +732,97 @@ def get_columns_in_relation(self, relation: AthenaRelation) -> List[AthenaColumn
...
+                drop_target_sql.strip(),
+                "\n\n-- Copy staging to target",
+                copy_to_target_sql.strip(),
+                "\n\n-- Drop staging table",
+                drop_staging_sql.strip(),
+            ]
+        )
+
+
+class SnapshotMigrationRequired(CompilationError):
```
Comment (Jrmyy) [inline dbt/adapters/athena/impl.py:827]: Maybe it is worth adding a file `exceptions.py`

**Review by Jrmyy — CHANGES_REQUESTED** [review]

File: dbt/adapters/athena/impl.py (hunk 13 above)
Comment (BrechtDeVlieger) [reply dbt/adapters/athena/impl.py:759]: It's a print on purpose. If I log it, then the messages are prefixed with timestamps. That makes it very difficult to copy/paste the query. So it is printed first, and then also logged at the end

Comment (Jrmyy) [reply dbt/adapters/athena/impl.py:760]: But if it in the exception, it should be easy to copy / paster it no ?
It is just that I find it weird to have this `print` but if you tell me there is no other way we will keep it like that.

Comment (nicor88) [reply dbt/adapters/athena/impl.py:760]: > It's a print on purpose. If I log it, then the messages are prefixed with timestamps. That makes it very difficult to copy/paste the query. So it is printed first, and then also logged at the end
that is a fair point, but at the same time the logged query will appear in the stdout when we run dbt models, if some people use json parsing that will fuck up the output.

Let's then please consider to remove it and replace with logger.debug to unify that approach on the all rest of the adapter. Consistency first :)

Comment (BrechtDeVlieger) [reply dbt/adapters/athena/impl.py:760]: Right. I'm fine with changing to proper logging output. I just noticed that even if I use proper logging, it doesn't add the time information upfront. It only does that for the exception message apparently. So, currently the query is printed twice: once in the final error message, and once in the print statement. How do you suggest to move forward?

- Do we want to keep the query in the final error? We can skip it there, because it's relatively hard to parse anyway. You need to remove the timestamp prefix from each line. With a bit of vim magic that goes quite fast, but I think most users will find this very annoying. Just to give you an example snippet:
  ```
  15:30:37    -- Backup original table
  15:30:37    create table "awsdatacatalog"."blabla"."cc_all_snapshot__dbt_tmp_migration_backup"
  15:30:37      with (
  15:30:37        table_type='hive',
  15:30:37        is_external=true,external_location='s3://some-bucket/tables/blabla/cc_all_snapshot__dbt_tmp_migration_backup/70d57c5a-ea3f-4542-9d13-1b214af36c9f',
  15:30:37        format='parquet'
  15:30:37      )
  15:30:37      as
  15:30:37        select * from "awsdatacatalog"."blabla"."cc_all_snapshot";
  ```
  - Do we want to keep the alternative logging statement? If I set it to debug, and don't show it in the error, then it will be very hard for people to understand what they need to do. If I log on info or warning, then the multiline output has no log prefixes. Not sure how json parsers deal with that tbh. 

I suggest the following: leave the query in the error output and replace the print statement with a debug statement. In the error output I'll then mention that the query is easier available when running in debug. Wdyt?

Comment (Jrmyy) [reply dbt/adapters/athena/impl.py:760]: It seems reasonable to me.
But the error is occurring right after the debug right ? 
So we might want to remove the query from the final error.

Comment (BrechtDeVlieger) [reply dbt/adapters/athena/impl.py:760]: But then it won't be visible without the debug run. That might be annoying for developers that can't directly access the dbt environment right?

Comment (nicor88) [reply dbt/adapters/athena/impl.py:760]: > I suggest the following: leave the query in the error output and replace the print statement with a debug statement. In the error output I'll then mention that the query is easier available when running in debug. Wdyt?
I think that is fine to use for this `logger.error` the most visible the better, maybe in the logger.error message we can write to switch to debug mode to have a more deeper look at the logs.

Comment (BrechtDeVlieger) [reply dbt/adapters/athena/impl.py:760]: So it seems that if we log a multiline statement, it will look like this:
```
00:00:00 line 1
line 2
line 3
```

But it seems like dbt captures the exception message if something fails, and reformats the message before printing, so that it looks like this:

```
00:00:00 line 1
00:00:00 line 2
00:00:00 line 3
```

I'm quite fond of this:
```
logger.error(msg)
raise SnapshotMigrationRequired("Look into 1.5 dbt docs for a complete migration procedure")
```

**Review by Jrmyy — APPROVED** [review]

**Review by nicor88 — APPROVED** [review]
